src/sql/parts/notebook/notebook.component.ts

```diff
@@ -331,6 +337,47 @@ export class NotebookComponent extends AngularDisposable implements OnInit, OnDe
 		}
 	}
 
+	// We can keep this method in common utilities
+	private getFilePathFromRecentWorkspace(untitledResource: URI): string {
+		let untitledFileName = untitledResource.path + '.' + DEFAULT_NOTEBOOK_FILETYPE;
+
+		let lastActiveFile = this.historyService.getLastActiveFile();
+		if (lastActiveFile) {
+			return URI.file(paths.join(paths.dirname(lastActiveFile.fsPath), untitledFileName)).fsPath;
+		}
+
+		let lastActiveFolder = this.historyService.getLastActiveWorkspaceRoot('file');
+		if (lastActiveFolder) {
+			return URI.file(paths.join(lastActiveFolder.fsPath, untitledFileName)).fsPath;
+		}
+		return untitledFileName;
+	}
+
+	promptForPath(defaultPath: string): TPromise<string> {
+		return this.windowService.showSaveDialog({ defaultPath });
+	}
+
+	// Entry point to save notebook
+	public async saveNotebook(): Promise<boolean> {
+		let self = this;
+		let notebookUri = this.notebookParams.notebookUri;
+		if (notebookUri.scheme === Schemas.untitled) {
+			let dialogPath = this.getFilePathFromRecentWorkspace(notebookUri);
+			return this.promptForPath(dialogPath).then(path => {
+				if (path) {
+					let newNotebookUri = URI.parse(path);
+					self._model.notebookUri = newNotebookUri;
```

Comment:
Sooo.... this needs to be updated in the notebookService for tracking. This is effectively a rename. We need to handle this
- In the notebookService
- We need to update the MainThreadNotebook / ExtHostNotebook classes to handle this. The state of the notebook (the kernel etc.) should be the same, but the notebookUri must be updated. This is vital - I'm guessing if you try to run the cells after saving this will fail.
- In the mainThreadNotebooksAndEditors, in terms of handling this event and sending to the extension host. Looking at how text editors do it, I suspect it's a "close and reopen" from their side
- in the extHostNotebooksAndEditors, handling the rename (or should be a no-op if that's not the case)

Reply:
As discussed offline - I believe we can do this incrementally. Let's put these changes in 2 separate PRs:
- 1 should send rename to the service and this should notify mainThreadNotebooksAndEditors listeners so API is "correct"
- 1 should handle rename in the notebook manager. This one may need more work & extension changes. Luckily based on our current design, we can get away with this not happening (in this release) since we don't allow rename of existing files on disk. We must fix for next release.